source/common/http/utility.cc

```diff
@@ -254,46 +254,72 @@ bool maybeAdjustForIpv6(absl::string_view absolute_url, uint64_t& offset, uint64
   return true;
 }
 
-absl::string_view parseCookie(absl::string_view cookie_value, absl::string_view key) {
-  // Split the cookie header into individual cookies.
-  for (const auto& s : StringUtil::splitToken(cookie_value, ";")) {
-    // Find the key part of the cookie (i.e. the name of the cookie).
-    size_t first_non_space = s.find_first_not_of(' ');
-    size_t equals_index = s.find('=');
-    if (equals_index == absl::string_view::npos) {
-      // The cookie is malformed if it does not have an `=`. Continue
-      // checking other cookies in this header.
-      continue;
-    }
-    absl::string_view k = s.substr(first_non_space, equals_index - first_non_space);
-    // If the key matches, parse the value from the rest of the cookie string.
-    if (k == key) {
+void forEachCookie(
+    const HeaderMap& headers, const LowerCaseString& cookie_header,
+    const std::function<bool(absl::string_view, absl::string_view)> cookie_consumer) {
+  const Http::HeaderMap::GetResult cookie_headers = headers.get(cookie_header);
+
+  for (size_t index = 0; index < cookie_headers.size(); index++) {
+    auto cookie_header_value = cookie_headers[index]->value().getStringView();
+
+    // Split the cookie header into individual cookies.
+    for (const auto& s : StringUtil::splitToken(cookie_header_value, ";")) {
+      // Find the key part of the cookie (i.e. the name of the cookie).
+      size_t first_non_space = s.find_first_not_of(' ');
+      size_t equals_index = s.find('=');
+      if (equals_index == absl::string_view::npos) {
+        // The cookie is malformed if it does not have an `=`. Continue
+        // checking other cookies in this header.
+        continue;
+      }
+      absl::string_view k = s.substr(first_non_space, equals_index - first_non_space);
       absl::string_view v = s.substr(equals_index + 1, s.size() - 1);
 
       // Cookie values may be wrapped in double quotes.
       // https://tools.ietf.org/html/rfc6265#section-4.1.1
       if (v.size() >= 2 && v.back() == '"' && v[0] == '"') {
         v = v.substr(1, v.size() - 2);
       }
-      return v;
+
+      bool continue_iteration = cookie_consumer(k, v);
+      if (!continue_iteration) {
+        return;
+      }
     }
   }
-  return EMPTY_STRING;
 }
 
 std::string parseCookie(const HeaderMap& headers, const std::string& key,
                         const LowerCaseString& cookie) {
-  const Http::HeaderMap::GetResult cookie_headers = headers.get(cookie);
+  std::string value;
 
-  for (size_t index = 0; index < cookie_headers.size(); index++) {
-    auto cookie_header_value = cookie_headers[index]->value().getStringView();
-    absl::string_view result = parseCookie(cookie_header_value, key);
-    if (!result.empty()) {
-      return std::string{result};
+  // Iterate over each cookie & return if its value is not empty.
+  forEachCookie(headers, cookie, [&key, &value](absl::string_view k, absl::string_view v) -> bool {
+    if (key == k && !v.empty()) {
```

Comment:
I guess this new code implements different behavior and theoretically may break certain setups. Consider the case of
```
Cookie: a=; b=1; a=2
Cookie: a=3
```
The old `parseCookie(headers, "a", "cookie")` would return `3`. The new code returns `2` if I'm not mistaken.

I think it makes sense to add a test for this case and make sure it passes for both versions.

Reply:
Ah yes, this new version is not equivalent to the old one. I'll try to make both the variants (single-cookie and map) compatible with the old behaviour.

Reply:
On a related note it appears that even the old behaviours haven't been consistent over time. 

For example if we use this new test as a reference:
```
TEST(HttpUtility, TestParseCookieDuplicates) {
  TestRequestHeaderMapImpl headers{
      {"someheader", "10.0.0.1"},
      {"cookie", "a=; b=1; a=2"},
      {"cookie", "a=3; b=2"}};

  EXPECT_EQ(Utility::parseCookieValue(headers, "a"), "3");
  EXPECT_EQ(Utility::parseCookieValue(headers, "b"), "1");
}
```

It fails before https://github.com/envoyproxy/envoy/pull/17560 landed (changed header iter. from reverse to natural order)
```
# git status
HEAD detached at b57827c20

# bazel test //test/common/http:utility_test
...
NFO: 12 processes: 1 internal, 11 darwin-sandbox.
INFO: Build completed, 1 test FAILED, 12 total actions
//test/common/http:utility_test                                          FAILED in 2.5s

# less /private/var/tmp/_bazel_shubhamp/306aac4fa57fcfade8ca38e576399a40/execroot/envoy/bazel-out/darwin-fastbuild/testlogs/test/common/http/utility_test/test.log
...
[ RUN      ] HttpUtility.TestParseCookieDuplicates
test/common/http/utility_test.cc:559: Failure
Expected equality of these values:
  Utility::parseCookieValue(headers, "b")
    Which is: "2"
  "1"
Stack trace:
  0x10b50f3aa: Envoy::Http::HttpUtility_TestParseCookieDuplicates_Test::TestBody()
  0x1148c1fc4: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x114897a2b: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x114897963: testing::Test::Run()
  0x114898977: testing::TestInfo::Run()
... Google Test internal frames ...
```

But passes on main:
```
# less less /private/var/tmp/_bazel_shubhamp/306aac4fa57fcfade8ca38e576399a40/execroot/envoy/bazel-out/darwin-fastbuild/testlogs/test/common/http/utility_test/test.log
...
[ RUN      ] HttpUtility.TestParseCookieDuplicates
[       OK ] HttpUtility.TestParseCookieDuplicates (0 ms)
```

Reply:
Behaviour on main (and the older behaviour) is a little inconsistent because it:
- picks the first matching cookie from a single `cookie` header (i.e. resolves `a`'s value to an empty string if the header value looks like: `"a=; b=1; a=2"`)
- but exhaustively searches remaining cookie headers until a non empty value is found.
And this ^ makes parsing out duplicate cookies spilt between multiple cookie headers fun.

Any suggestions on how to proceed here:
1. Keep the behaviour from `main` as-is
2. Pick the `first cookie` OR `first non-empty cookie`
3. Pick the `last cookie` OR `last non-empty cookie`

I'm inclined towards approaches 2 or 3 because it keeps the parsing simple in the long run - but this can break some setups.

[RFC 6265 - HTTP State Management Mechanism](https://datatracker.ietf.org/doc/html/rfc6265#section-4.2.2) itself is pretty ambiguous about duplicate cookies and ordering:
```
   Although cookies are serialized linearly in the Cookie header,
   servers SHOULD NOT rely upon the serialization order.  In particular,
   if the Cookie header contains two cookies with the same name (e.g.,
   that were set with different Path or Domain attributes), servers
   SHOULD NOT rely upon the order in which these cookies appear in the
   header.
```

Reply:
Any suggestions on how to proceed here:
1. Keep the behaviour from `main` as-is
2. Pick the `first cookie` OR `first non-empty cookie`
3. Pick the `last cookie` OR `last non-empty cookie`

I'm inclined towards approaches 2 or 3 because it keeps the parsing simple in the long run - but this can break some setups.

[RFC 6265 - HTTP State Management Mechanism](https://datatracker.ietf.org/doc/html/rfc6265#section-4.2.2) itself is pretty ambiguous about duplicate cookies and ordering:
```
   Although cookies are serialized linearly in the Cookie header,
   servers SHOULD NOT rely upon the serialization order.  In particular,
   if the Cookie header contains two cookies with the same name (e.g.,
   that were set with different Path or Domain attributes), servers
   SHOULD NOT rely upon the order in which these cookies appear in the
   header.
```

Reply:
Hm... That's a good catch! If the spec explicitly states that the order is unimportant then just pick the first value irrespective of its emptiness for the sake of perf. 

A second reviewer may give a second opinion on it.